File: backend/app/integrations/ai/azure_openai_client.py
import json
from typing import Any, Dict, List

from openai import AzureOpenAI
from app.core.config import settings


def _client() -> AzureOpenAI:
    if not settings.azure_openai_endpoint or not settings.azure_openai_key:
        raise ValueError('Azure OpenAI endpoint/key not configured')
    return AzureOpenAI(azure_endpoint=settings.azure_openai_endpoint, api_key=settings.azure_openai_key, api_version=settings.azure_openai_api_version)


def chat_completion_json(system_prompt: str, user_prompt: str) -> Dict[str, Any]:
    c = _client()
    if not settings.azure_openai_deployment:
        raise ValueError('AZURE_OPENAI_DEPLOYMENT not set')
    resp = c.chat.completions.create(
        model=settings.azure_openai_deployment,
        messages=[{'role':'system','content':system_prompt},{'role':'user','content':user_prompt}],
        temperature=settings.llm_temperature,
    )
    content = (resp.choices[0].message.content or '').strip()
    if content.startswith('```'):
        content = content.strip('`').replace('json', '')
    return json.loads(content)


def embed_texts(texts: List[str]) -> List[List[float]]:
    c = _client()
    if not settings.azure_openai_embedding_deployment:
        raise ValueError('AZURE_OPENAI_EMBEDDING_DEPLOYMENT not set')
    r = c.embeddings.create(model=settings.azure_openai_embedding_deployment, input=texts)
    return [d.embedding for d in r.data]

File: backend/src/integrations/ai/azure-openai-client.ts
import { settings } from '../../core/settings.js';
import crypto from 'crypto';

function isAzureConfigured(): boolean {
  return !!(settings.azureOpenaiEndpoint && settings.azureOpenaiKey);
}

async function getClient() {
  const { AzureOpenAI } = await import('openai');
  if (!settings.azureOpenaiEndpoint || !settings.azureOpenaiKey) {
    throw new Error('Azure OpenAI endpoint/key not configured');
  }
  return new AzureOpenAI({
    endpoint: settings.azureOpenaiEndpoint,
    apiKey: settings.azureOpenaiKey,
    apiVersion: settings.azureOpenaiApiVersion,
  });
}

/**
 * Chat completion that returns the raw string response.
 * Used by RAG answer generation where we want natural language, not JSON.
 */
export async function chatCompletion(
  systemPrompt: string,
  userPrompt: string
): Promise<string> {
  const client = await getClient();
  if (!settings.azureOpenaiDeployment) {
    throw new Error('AZURE_OPENAI_DEPLOYMENT not set');
  }
  const resp = await client.chat.completions.create({
    model: settings.azureOpenaiDeployment,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt },
    ],
    temperature: settings.llmTemperature,
  });
  return (resp.choices[0]?.message?.content || '').trim();
}

/**
 * Chat completion that returns parsed JSON.
 */
export async function chatCompletionJson(
  systemPrompt: string,
  userPrompt: string
): Promise<Record<string, unknown>> {
  const client = await getClient();
  if (!settings.azureOpenaiDeployment) {
    throw new Error('AZURE_OPENAI_DEPLOYMENT not set');
  }
  const resp = await client.chat.completions.create({
    model: settings.azureOpenaiDeployment,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt },
    ],
    temperature: settings.llmTemperature,
  });
  let content = (resp.choices[0]?.message?.content || '').trim();
  if (content.startsWith('```')) {
    content = content.replace(/^```json?\n?/, '').replace(/```$/, '').trim();
  }
  return JSON.parse(content);
}

/**
 * Generate a deterministic local embedding from text using SHA-256.
 * Produces a 256-dimensional unit vector. Not semantically meaningful,
 * but allows the full RAG pipeline to run without Azure OpenAI.
 */
function localEmbed(text: string): number[] {
  const hash = crypto.createHash('sha256').update(text).digest();
  const vec: number[] = [];
  for (let i = 0; i < hash.length; i++) {
    // Map byte [0,255] -> [-1,1]
    vec.push((hash[i] / 127.5) - 1);
  }
  // normalize to unit vector
  let norm = 0;
  for (const v of vec) norm += v * v;
  norm = Math.sqrt(norm);
  if (norm > 0) {
    for (let i = 0; i < vec.length; i++) vec[i] /= norm;
  }
  return vec;
}

/**
 * Generate embeddings for texts.
 * Falls back to a local hash-based embedding when Azure OpenAI is not configured.
 */
export async function embedTexts(texts: string[]): Promise<number[][]> {
  if (!isAzureConfigured()) {
    console.log(`[RAG] Using local hash embeddings (Azure OpenAI not configured) for ${texts.length} text(s)`);
    return texts.map(t => localEmbed(t));
  }

  const client = await getClient();
  if (!settings.azureOpenaiEmbeddingDeployment) {
    throw new Error('AZURE_OPENAI_EMBEDDING_DEPLOYMENT not set');
  }
  const r = await client.embeddings.create({
    model: settings.azureOpenaiEmbeddingDeployment,
    input: texts,
  });
  return r.data.map(d => d.embedding);
}
